src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -2154,22 +2290,37 @@ private MethodSymbol InferMethod(BoundCall node, MethodSymbol method, ImmutableA
             return method;
         }
 
-        private static readonly Func<TypeSymbolWithAnnotations, BoundExpression, BoundExpression> s_makePlaceholderForArgumentFunc =
-            (TypeSymbolWithAnnotations argumentType, BoundExpression argument) =>
+        private ImmutableArray<BoundExpression> GetArgumentsForMethodTypeInference(ImmutableArray<BoundExpression> arguments, ImmutableArray<Result> argumentResults)
+        {
+            // PROTOTYPE(NullableReferenceTypes): MethodTypeInferrer.Infer relies
+            // on the BoundExpressions for tuple element types and method groups.
+            // By using a generic BoundValuePlaceholder, we're losing inference in those cases.
+            // PROTOTYPE(NullableReferenceTypes): Inference should be based on
+            // unconverted arguments. Consider cases such as `default`, lambdas, tuples.
+            return arguments.ZipAsArray(argumentResults, (argument, result) => getArgumentForMethodTypeInference(argument, result));
```

Comment:
I suspect a new delegate is created on every call here. If that's the case, it's probably good to cache (like before).

Reply:
Yes, a new delegate is created for each call. Unfortunately, the local function requires `this` now so at the very least we'll need an instance field to cache this. But that instance field would need to be set lazily to allow access to `this`. Will leave as is for now rather than introducing too much code to work around the delegate.

---
In reply to: [194541160](https://github.com/dotnet/roslyn/pull/27283#discussion_r194541160) [](ancestors = 194541160)